config/filter.d/bitwarden.conf

```diff
@@ -0,0 +1,9 @@
+# Fail2Ban filter for Bitwarden
+# Detecting failed login attempts
+# Logged in bwdata/logs/identity/Identity/log.txt
+
+[Init]
+datepattern = ^%%Y-%%m-%%d %%H:%%M:%%S
+
+[Definition]
+failregex = ^.*Failed login attempt(, 2FA invalid)?\. <HOST>$
```

Comment:
Regex `^.*` is not really an anchor (it works same way as it were not there). And after cut out of string matched default datepattern it has pretty simple prefix, so better would be to use something like this:
```diff
- failregex = ^.*Failed login attempt(, 2FA invalid)?\. <HOST>$
+ failregex = ^\s*\[WRN\]\s+Failed login attempt(?:, 2FA invalid)?\. <HOST>$
```
Also note it uses non capturing group `(?: ...)`.

Reply:
Perfect, many thanks for your review and feedback.
In the updated the rule accordingly, with just a little modification :
`failregex = ^\s*\[\w+\]\s+Failed login attempt(?:, 2FA invalid)?\. <HOST>$`
As I'm not sure `WRN` will remain, but for sure we'll have a word there.

Reply:
> As I'm not sure `WRN` will remain

Hmm... why it should not? The same would be valid if `[WRN]` get rewritten without brackets, or completely removed together with brackets. Or if IP will be moved to begin of log.
You cannot absolutely protect the filter against any log-format modifications that devs could do at some point...
Anyway having WRT could speedup it a bit (for example if some other messages logged also, e. g. with something like `[INFO]` or so).
I'm for WRN... But you could define it as parameter:
```ini
[Definition]
failloglev = WRN
failregex = ^\s*\[%(failloglev)s\]\s+Failed login attempt(?:, 2FA invalid)?\. <HOST>$
```
so it'd be possible to overwrite it in jail.local:
```ini
[bitwarden]
filter = bitwarden[failloglev='\w+']
```

Reply:
Let's then get back to `WRN`, the easiest solution, I'm OK with this :)
We'll update the jail whenever log format changes.
Thank you again 👍